File: backend/app/schemas/client_experience.py
from __future__ import annotations

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.financial_analysis import FinancialAnalysisResponse
from app.schemas.planning import PlanningResponse
from app.schemas.trust import KnowledgeCitation


class ClientJourneyStep(BaseModel):
    code: str
    label: str
    status: Literal["completed", "ready", "needs_action", "blocked"]
    reason: str
    action: str
    view: Literal[
        "family",
        "balance",
        "cashflow",
        "health",
        "accounts",
        "goals",
        "twin",
        "behavior",
        "report",
        "actions",
        "privacy",
    ]


class PrivacyConsentItem(BaseModel):
    id: str
    scopes: list[str]
    purpose: str
    granted_at: datetime
    withdrawn_at: datetime | None
    consent_version: str
    record_version: int
    status: Literal["active", "withdrawn"]
    withdrawal_allowed: bool
    scenario: str
    sensitive: bool
    explicit: bool
    minimum_necessary: bool


class ClientPrivacySummary(BaseModel):
    active_consent_count: int
    withdrawn_consent_count: int
    consents: list[PrivacyConsentItem]
    export_path: str
    delete_path: str
    deletion_mode: Literal["logical_erasure_with_confirmation"] = (
        "logical_erasure_with_confirmation"
    )
    human_review_path: str
    boundary_note: str


class ActionCalendarItem(BaseModel):
    code: str
    title: str
    detail: str
    why: str
    constraint_or_formula: str
    change_trigger: str
    risk_and_assumptions: str
    amount: str
    due_date: date | None
    priority: int
    source_record_ids: list[str]
    calculation_source: Literal["deterministic_planning_rules", "deterministic_review_schedule"]


class ActionCalendarGroup(BaseModel):
    code: Literal["immediate", "three_months", "one_year", "long_term", "next_12_months"]
    label: str
    description: str
    items: list[ActionCalendarItem]


class ClientReportChapter(BaseModel):
    number: int = Field(ge=1, le=8)
    title: str
    summary: str
    calculation_basis: list[str]
    citation_ids: list[str]
    status: Literal["ready", "pending_twin", "needs_review"]


class ClientReportPreview(BaseModel):
    report_version: str
    knowledge_retrieval_version: str
    title: str
    subtitle: str
    chapter_count: Literal[8] = 8
    chapters: list[ClientReportChapter] = Field(min_length=8, max_length=8)
    citations: list[KnowledgeCitation]
    generated_at: datetime
    data_as_of: date
    calculation_source: Literal["deterministic_client_experience_composer"] = (
        "deterministic_client_experience_composer"
    )
    boundary_note: str


class ClientDeliveryState(BaseModel):
    report: Literal["not_generated", "under_review", "client_ready"]
    workflow: Literal["not_created", "under_review", "client_ready"]
    actions: Literal["not_generated", "under_review", "client_ready"]
    explanation: str


class ClientExperienceResponse(BaseModel):
    household_id: str
    household_code: str
    household_name: str
    household_version: int
    synthetic_data: bool
    analysis_date: date
    data_as_of: date
    journey: list[ClientJourneyStep]
    privacy: ClientPrivacySummary
    action_calendar: list[ActionCalendarGroup]
    report: ClientReportPreview
    delivery: ClientDeliveryState
    calculation_versions: dict[str, str]
    state_catalog: list[str]
    mock_mode_supported: Literal[True] = True


class ClientDataExport(BaseModel):
    package_version: str
    exported_at: datetime
    household_id: str
    financial_analysis: FinancialAnalysisResponse
    planning: PlanningResponse
    client_experience: ClientExperienceResponse
    boundary_note: str


class ConsentWithdrawRequest(BaseModel):
    expected_version: int = Field(ge=1)
    reason: str = Field(min_length=2, max_length=300)


class HumanReviewRequest(BaseModel):
    reason: str = Field(min_length=2, max_length=500)
    context: str = Field(default="client_privacy_center", min_length=2, max_length=120)


class HumanReviewResponse(BaseModel):
    request_id: str
    household_id: str
    status: Literal["queued"] = "queued"
    queue: Literal["demo_advisor_queue"] = "demo_advisor_queue"
    created_at: datetime
    message: str

File: frontend/src/api/clientExperience.ts
import type { KnowledgeCitation } from "./trust";

export type ClientView =
  | "family"
  | "balance"
  | "cashflow"
  | "health"
  | "accounts"
  | "pension"
  | "goals"
  | "twin"
  | "behavior"
  | "report"
  | "actions"
  | "privacy";

export interface ClientJourneyStep {
  code: string;
  label: string;
  status: "completed" | "ready" | "needs_action" | "blocked";
  reason: string;
  action: string;
  view: ClientView;
}

export interface PrivacyConsent {
  id: string;
  scopes: string[];
  purpose: string;
  granted_at: string;
  withdrawn_at: string | null;
  consent_version: string;
  record_version: number;
  status: "active" | "withdrawn";
  withdrawal_allowed: boolean;
  scenario: string;
  sensitive: boolean;
  explicit: boolean;
  minimum_necessary: boolean;
}

export interface ClientPrivacySummary {
  active_consent_count: number;
  withdrawn_consent_count: number;
  consents: PrivacyConsent[];
  export_path: string;
  delete_path: string;
  deletion_mode: "logical_erasure_with_confirmation";
  human_review_path: string;
  boundary_note: string;
}

export interface ActionCalendarItem {
  code: string;
  title: string;
  detail: string;
  why: string;
  constraint_or_formula: string;
  change_trigger: string;
  risk_and_assumptions: string;
  amount: string;
  due_date: string | null;
  priority: number;
  source_record_ids: string[];
  calculation_source: "deterministic_planning_rules" | "deterministic_review_schedule";
}

export interface ActionCalendarGroup {
  code: "immediate" | "three_months" | "one_year" | "long_term" | "next_12_months";
  label: string;
  description: string;
  items: ActionCalendarItem[];
}

export interface ClientReportChapter {
  number: number;
  title: string;
  summary: string;
  calculation_basis: string[];
  citation_ids: string[];
  status: "ready" | "pending_twin" | "needs_review";
}

export interface ClientReportPreview {
  report_version: string;
  knowledge_retrieval_version: string;
  title: string;
  subtitle: string;
  chapter_count: 8;
  chapters: ClientReportChapter[];
  citations: KnowledgeCitation[];
  generated_at: string;
  data_as_of: string;
  calculation_source: "deterministic_client_experience_composer";
  boundary_note: string;
}

export interface ClientDeliveryState {
  report: "not_generated" | "under_review" | "client_ready";
  workflow: "not_created" | "under_review" | "client_ready";
  actions: "not_generated" | "under_review" | "client_ready";
  explanation: string;
}

export interface ClientExperience {
  household_id: string;
  household_code: string;
  household_name: string;
  household_version: number;
  synthetic_data: boolean;
  analysis_date: string;
  data_as_of: string;
  journey: ClientJourneyStep[];
  privacy: ClientPrivacySummary;
  action_calendar: ActionCalendarGroup[];
  report: ClientReportPreview;
  delivery: ClientDeliveryState;
  calculation_versions: Record<string, string>;
  state_catalog: string[];
  mock_mode_supported: true;
}

export interface HumanReviewResponse {
  request_id: string;
  household_id: string;
  status: "queued";
  queue: "demo_advisor_queue";
  created_at: string;
  message: string;
}

export interface PrivacyRequestResult {
  request_id: string;
  request_type: "export" | "erase";
  status: "completed" | "rejected";
  result_summary: Record<string, unknown>;
}

export interface PrivacyExportResponse {
  request: PrivacyRequestResult;
  package_version: string;
  exported_at: string;
  household_ref: string;
  data: Record<string, unknown>;
}

const apiBase = (import.meta.env.VITE_API_BASE_URL ?? "").replace(/\/$/, "");

async function readJson<T>(path: string, signal?: AbortSignal, init?: RequestInit): Promise<T> {
  const response = await fetch(apiBase + path, {
    ...init,
    signal,
    headers: { Accept: "application/json", ...init?.headers },
  });
  if (!response.ok) throw new Error(`API ${response.status}: ${path}`);
  return (await response.json()) as T;
}

export function fetchClientExperience(
  householdId: string,
  signal?: AbortSignal,
): Promise<ClientExperience> {
  return readJson(
    `/api/v1/households/${encodeURIComponent(householdId)}/client-experience`,
    signal,
  );
}

export function clientExperienceExportUrl(householdId: string): string {
  return `${apiBase}/api/v1/households/${encodeURIComponent(householdId)}/client-experience/export`;
}

export async function exportHouseholdData(householdId: string): Promise<void> {
  const result = await readJson<PrivacyExportResponse>(
    `/api/v1/households/${encodeURIComponent(householdId)}/privacy/exports`,
    undefined,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-Confirm-Action": "export_household_data",
      },
      body: JSON.stringify({ reason: "客户在隐私中心主动导出家庭数据" }),
    },
  );
  const blobUrl = URL.createObjectURL(
    new Blob([JSON.stringify(result, null, 2)], { type: "application/json" }),
  );
  const link = document.createElement("a");
  link.href = blobUrl;
  link.download = `wealthtwin-privacy-export-${result.household_ref}.json`;
  link.click();
  URL.revokeObjectURL(blobUrl);
}

export function withdrawPrivacyConsent(
  householdId: string,
  consentId: string,
  expectedVersion: number,
  reason: string,
): Promise<PrivacyConsent> {
  return readJson(
    `/api/v1/households/${encodeURIComponent(householdId)}/privacy/consents/${encodeURIComponent(consentId)}/withdraw`,
    undefined,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ expected_version: expectedVersion, reason }),
    },
  );
}

export function requestHumanReview(
  householdId: string,
  reason: string,
): Promise<HumanReviewResponse> {
  return readJson(
    `/api/v1/households/${encodeURIComponent(householdId)}/privacy/human-review-requests`,
    undefined,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ reason, context: "client_privacy_center" }),
    },
  );
}

export async function deleteHouseholdData(
  householdId: string,
  expectedVersion: number,
  householdCode: string,
  reason: string,
): Promise<PrivacyRequestResult> {
  return readJson<PrivacyRequestResult>(
    `/api/v1/households/${encodeURIComponent(householdId)}/privacy/deletion-requests`,
    undefined,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-Confirm-Action": "erase_household_data",
      },
      body: JSON.stringify({
        expected_version: expectedVersion,
        household_code_confirmation: householdCode,
        reason,
      }),
    },
  );
}
